src/lib.rs

```diff
@@ -591,44 +600,18 @@ mod tests {
         s
     }
 
-    // In cases where all we have is a bom, we expect the bytes to be
-    // passed through unchanged.
-    #[test]
```

Comment:
This requires some explanation.

The BOM is not random bytes, it is a unicode character, which is decodable just like any other unicode character. The bytes that comprise the BOM are merely the representation of that unicode character in a particular encoding.

You never, ever want to just "pass the bytes through unchanged".

Reply:
I understand everything you've written here, but don't understand why that has led you to remove the test. Could you say more please?

Reply:
The tests started failing as soon as I fixed the bug here https://github.com/BurntSushi/encoding_rs_io/pull/11#discussion_r962187963

It was returning the bytes unchanged was because it wasn't detecting an encoding. Once the encoding was detected, it starts sending bytes through the decoder, which strips the BOM because `encoding.new_decoder_with_bom_removal()` is used.  So the tests failed because the output was empty.

What probably ought to be done instead of just removing them entirely is - fix the decoding part as described, so that the BOM is not removed by the decoder, and then have these tests ensure that a BOM is returned in all cases where present in the input except when the user explicitly asked for its removal via `strip_bom`.

Reply:
I still don't get it. I'm going to have to go through and recontexualize the library impl in order to accept this. I don't know when I'll get the chance to do that.

Reply:
Ok.  If it helps, I can walk through my own steps.

* Check out the first commit ([16b3203](https://github.com/BurntSushi/encoding_rs_io/pull/11/commits/16b320367f700061156361a56dc4ca6b94db3801)) and add these lines back https://github.com/BurntSushi/encoding_rs_io/pull/11#discussion_r962187963
* Observe that the new `PossibleBom::encoding()` tests fail by returning `None` when they ought to recognize a UTF-16 BOM and return `Some($encoding)`
* Review the tests for correctness.  Reset back to the original commit.  Observe that with the original commit, now the `trans_utf16_bom` tests are failing

```
---- tests::trans_utf16_bom stdout ----
thread 'tests::trans_utf16_bom' panicked at 'assertion failed: `(left == right)`
  left: `[255, 254]`,
 right: `[]`', src/lib.rs:602:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


failures:
    tests::trans_utf16_bom

test result: FAILED. 36 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
```

* After debugging it becomes clear that the reason was because the decoder was swallowing the BOM even though `DecodeReaderBytes` was not configured to strip the BOM.  And that is because we're creating the decoder with `encoding.new_decoder_with_bom_removal()`

Reply:
The actual behavior it should probably expect, is not to get either an empty buffer or `[FF, FE]`, but to get `[0xEF, 0xBB, 0xBF]`, the BOM character decoded to UTF-8.

Reply:
@BurntSushi As a practical matter, I need to know if this can possibly be merged soon or if I should just vendor the code.  Will you get an opportunity to review soon?

Reply:
@BurntSushi ^

Reply:
Is this change an extension of https://github.com/hsivonen/encoding_rs/issues/88? Because I generally agree with @hsivonen there. Otherwise, I'm still going to have to re-contextualize myself, and as I already said, **I don't know** when I'll be able to do this. Basically, what you're saying just doesn't make sense to me. That doesn't mean you're wrong. What I know is that this crate's implementation is rather subtle and I am just not okay with merging things I don't understand.

Basically, I wrote this test for _some_ reason, so if we're going to change the code in a way that results in the test failing, then I personally need to be exceptionally clear that it's indeed the right thing to do. And that it isn't going to break downstream folks. Including me. A lot of this crate was written for use in ripgrep for example.